### ISSUE TITLE
`DokkatooGenerateTask.dokkaParametersJson` contains absolute paths, making the task non-relocatable

### ISSUE THREAD

**Issue by martinbonnin** [adamko-dev/dokkatoo#66]
[Build scan comparisons](https://ge.apollographql.com/c/q7k3uzw5otxyy/unxcnayzprq56/task-inputs?cacheability=cacheable,overlapping-outputs,validation-failure&expanded=WyI1Mnh0ZGNkb2s2dGhnLWRva2thbW9kdWxlZmlsZXMiLCJiY2M2NHNtYjNzYnVpLWRva2thc291cmNlc2V0cy5jb21tb25tYWluJDIuY2xhc3NwYXRoIl0#ee7filp2swu5e-dokkaparametersjson)

Example on my machine: 

```
$ cat libraries/apollo-adapters/build/dokka-config/html/dokka_parameters.json | head -n 20
{
  "moduleName": "apollo-adapters",
  "moduleVersion": "4.0.0-dev.3-SNAPSHOT",
  "offlineMode": false,
  "failOnWarning": false,
  "sourceSets": [
    {
      "sourceSetId": {
        "scopeId": ":apollo-adapters",
        "sourceSetName": "appleMain"
      },
      "displayName": "apple",
      "classpath": [
        "/Users/mbonnin/git/apollo-kotlin/libraries/apollo-api/build/classes/kotlin/iosArm64/main/klib/apollo-api.klib",
        "/Users/mbonnin/git/apollo-kotlin/libraries/apollo-annotations/build/classes/kotlin/iosArm64/main/klib/apollo-annotations.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlinx/kotlinx-datetime-iosarm64/0.4.0/8f2e032c52e3e59cf8e63213609007f80bdb6748/kotlinx-datetime.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/com.squareup.okio/okio-iosarm64/3.2.0/a4351920c1dc049812599064a73596d245aa12c7/okio.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/com.benasher44/uuid-iosarm64/0.6.0/719bcb24170ab712e6b31de02a6d6ad863bc926c/uuid.klib",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.6.21/11ef67f1900634fd951bad28c53ec957fabbe5b8/kotlin-stdlib-1.6.21.jar",
        "/Users/mbonnin/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/24.0.1/13c5c75c4206580aa4d683bffee658caae6c9f43/annotations-24.0.1.jar",
```

Not really sure how to make the `/Users/mbonnin/...` relative. Maybe the project dependencies can be relative to the root project but the external ones sound tricky

**Comment by aSemy** [comment]
I wondered how long it would take for you to find this problem 😄. It's a big shortcoming with the Dokka configuration. I'm stumped as to how to fix it in a nice way. 

The only idea I have so far is to try and do some clever find-replace to try and replace files that start with `$GRADLE_USER_HOME` with a `GRADLE_USER_HOME` string (and similar for `rootProjectDir`), and try to convert it on read and write... but that seems too complicated.

It might be possible for the classpath to just not be serialized to JSON, because it's passed around using a configuration. The classpath should then be an input for the `DokkaGeneratorWorker`, which would create a final `DokkaParameters` instance that is passed to the actual Dokka Generator.

BUT it's a problem for all `File` properties in https://github.com/adamko-dev/dokkatoo/blob/v1.3.0/modules/dokkatoo-plugin/src/main/kotlin/dokka/parameters/DokkaParametersKxs.kt. Most of those files aren't defined in a Configuration.

Maybe it's possible to apply some [custom Gradle normalization](https://docs.gradle.org/current/userguide/incremental_build.html#sec:configure_input_normalization)? https://github.com/gradle/gradle/issues/16225

**Comment by aSemy** [comment]
Alright, I think this is the trick:

* Instead of marking the JSON files as `@Input`, instead mark them as `@Internal`, so Gradle ignores them
* Create utility functions for converting Publication/Module descriptor JSON files to `DokkaParametersSpec`/`DokkaModuleDescriptionSpec`
* Create a new `@Nested` properties that provides the Gradle types, `Provider<DokkaParametersSpec>`, `Provider<List<DokkaModuleDescriptionSpec>>`

Converting the JSON back to Gradle-managed types means Gradle's normalization annotations can work.

This approach also helps with aligning the task properties in both the 'prepare parameters' and 'generator' tasks.

Writing the parameter JSONs to file is still necessary for aggregating subprojects, as there is no other way for Gradle tasks to share output values other than via a file.

**Comment by aSemy** [comment]
I've been stuck on this for two weeks, and I've reached the point where I don't think this is possible to achieve in Gradle. Sharing files between subprojects is too flakey and poorly documented, with opaque APIs and errors. I've not seen any improvement on this in 2 years https://github.com/gradle/gradle/issues/18846.

* Attributes are randomly ignored, so non-matching files get added to configurations. Apparently this is intentional https://github.com/gradle/gradle/issues/23776 which is completely incomprehensible to me
* There's no guide on the best way to resolve files https://github.com/gradle/gradle/issues/23772 and the Gradle team doesn't want one
* There's no documentation on how to use variants to share files
* There's no documentation on the `artifactType` attribute
* I've tried using `AttributeCompatibilityRule` and `ExactMatchDisambiguationRule` rules, but Gradle ignores them. Was I using them incorrectly, or do they just not work? It's a mystery.
